File: src/main/java/dev/esophose/discordbot/command/commands/moderation/MuteCommand.java
package dev.esophose.discordbot.command.commands.moderation;

import dev.esophose.discordbot.Sparky;
import dev.esophose.discordbot.command.DiscordCommand;
import dev.esophose.discordbot.command.DiscordCommandMessage;
import dev.esophose.discordbot.manager.CommandManager;
import discord4j.core.object.entity.Guild;
import discord4j.core.object.entity.Member;
import discord4j.core.object.entity.Role;
import discord4j.core.object.util.Permission;
import discord4j.core.object.util.PermissionSet;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import org.apache.commons.lang3.StringUtils;

public class MuteCommand extends DiscordCommand  {

    public MuteCommand(Sparky bot) {
        super(bot);
    }

    public void execute(DiscordCommandMessage message, Member member, Optional<String> reason) {
        CommandManager commandManager = this.bot.getManager(CommandManager.class);
        message.getGuild().flatMapMany(Guild::getRoles).collectList().subscribe(roles -> {
            Optional<Role> optionalMutedRole = roles.stream().filter(x -> StringUtils.equalsIgnoreCase(x.getName(), "Muted")).findFirst();
            if (!optionalMutedRole.isPresent()) {
                commandManager.sendResponse(message.getChannel(), "Not set up", "The mute command is not yet set up. Please run `.setupmute` to set up the role and channel overrides.").subscribe();
            } else {
                member.addRole(optionalMutedRole.get().getId(), reason.orElse("Member was muted")).subscribe();
                commandManager.sendResponse(message.getChannel(), "Member muted", member.getNicknameMention() + " has been muted.").subscribe();
            }
        });
    }

    @Override
    public String getName() {
        return "mute";
    }

    @Override
    public List<String> getAliases() {
        return Collections.emptyList();
    }

    @Override
    public String getDescription() {
        return "Mutes a member";
    }

    @Override
    public PermissionSet getRequiredBotPermissions() {
        return PermissionSet.of(Permission.SEND_MESSAGES, Permission.MANAGE_ROLES, Permission.MUTE_MEMBERS);
    }

    @Override
    public Permission getDefaultRequiredMemberPermission() {
        return Permission.MUTE_MEMBERS;
    }

}

File: src/main/kotlin/dev/esophose/discordbot/command/commands/moderation/MuteCommand.kt
package dev.esophose.discordbot.command.commands.moderation

import dev.esophose.discordbot.Sparky
import dev.esophose.discordbot.command.DiscordCommand
import dev.esophose.discordbot.command.DiscordCommandMessage
import dev.esophose.discordbot.manager.CommandManager
import discord4j.core.`object`.entity.Member
import discord4j.rest.util.Permission
import discord4j.rest.util.PermissionSet
import org.apache.commons.lang3.StringUtils
import java.util.*

class MuteCommand : DiscordCommand() {

    override val name: String
        get() = "mute"

    override val aliases: List<String>
        get() = emptyList()

    override val description: String
        get() = "Mutes a member"

    override val requiredBotPermissions: PermissionSet
        get() = PermissionSet.of(Permission.SEND_MESSAGES, Permission.MANAGE_ROLES, Permission.MUTE_MEMBERS)

    override val defaultRequiredMemberPermission: Permission
        get() = Permission.MUTE_MEMBERS

    fun execute(message: DiscordCommandMessage, member: Member, reason: Optional<String>) {
        val commandManager = Sparky.getManager(CommandManager::class)

        if (member.id == Sparky.self.id) {
            commandManager.sendResponse(message.channel, "Why would you do that?", "Why would you try to mute me? What have I ever done to you?").subscribe()
            return
        }

        message.guild.flatMapMany { it.roles }.collectList().subscribe { roles ->
            val optionalMutedRole = roles.stream().filter { x -> StringUtils.equalsIgnoreCase(x.name, "Muted") }.findFirst()
            if (!optionalMutedRole.isPresent) {
                commandManager.sendResponse(message.channel, "Not set up", "The mute command is not yet set up. Please run `.setupmute` to set up the role and channel overrides.").subscribe()
            } else {
                member.addRole(optionalMutedRole.get().id, reason.orElse("Member was muted")).subscribe()
                commandManager.sendResponse(message.channel, "Member muted", member.nicknameMention + " has been muted.").subscribe()
            }
        }
    }

}
